### ISSUE TITLE
Refactor Volume Code

### ISSUE THREAD

**Issue by saad-ali** [kubernetes/kubernetes#14217]
The `Volume` interface defined in `pkg/volume/volume.go` defines methods that are used for mounting/unmounting/attaching/detaching but also volume creation/deletion. The former is used by the kubelet binary on the node, and the later by the persistent volume controller on the master.

While we should keep all volume related code in the same directory, we should consider refactoring to separate kubelet only/master only code into different packages and move common code into a shared dependent package.

**Comment by thockin** [comment]
I like this idea.

**Comment by markt-gh** [comment]
All provisioning stuff is only marginally related to volumes. They act more
on API objects than volume behavior (mounting, attaching, whatever).

Considering we want to encourage swapping out controllers and provisioning
by users, should all that kind of code be packaged more like the scheduler
plugin in the /plugins?  We can focus on making them easier to change and
configure.

**Comment by thockin** [comment]
I think there's value to having all of the per-volume-kind code in one
place, even if some of it is orchestration-related and some of it is
node-related.  But breaking into multiple pkgs doesn't sound terrible.

**Comment by markturansky** [comment]
Can it be as simple as moving all the provisioning/recycling related stuff to provisioning.go within the existing package structure?  It would just be cut and paste from one file to another. No import changes anywhere.

```
pkg/volume/
  aws_ebs/
    aws_ebs.go
    aws_ebs_test.go
    provisioning.go
    provisioning_test.go
  gce_pd/
    gce_pd.go
    provisioning.go
    ...
```

**Comment by markturansky** [comment]
@pmorie see above comment

**Comment by saad-ali** [comment]
I think it'll be bigger than that: `volume.VolumePluginMgr` which is required to `InitPlugins` requires a `VolumeHost` object. The `VolumeHost` object is very kubelet specific (gets `KubeClient`, `VolumeDir`, etc.). So non-kubelet packages [currently stub out](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go#L486) that functionality to make it work; this needs to be fixed.

**Comment by markturansky** [comment]
Agreed that we should make an appropriate `VolumeHost` for non-mounting plugins.  

The controller is just one.  Jan is adding plugins to the API server to fix an Admission issue where an admission plugin looks-up cloud labels and applies them to a PV.  I believe this one is like the controller in that it only needs the cloud provider.

I think these are at least 2 PRs.  One for an appropriate host impl and another for separating provisioning logic into separate files.  Arguably separate PRs for each plugin being refactored.

**Comment by jsafrane** [comment]
I started splitting provisioners from generic volume plugins in https://github.com/kubernetes/kubernetes/pull/21559. I needed to create a special `MountableVolumePlugin` interface - until now, all volume plugins needed to implement `NewBuilder`/`NewCleaner` interfaces, with the PR it's only optional. And cloud provisioners won't implement it - there will be eventually many of them (AWS SSD / AWS magnetic / several IOPS classes), while there will be one mounter/attacher/deleter. These can be split into separate files / objects too, if we need to.

Splitting `VolumeHost` into kubelet/apiserver parts is the next logical step, I'm talking to @swagiaal about it.

**Comment by saad-ali** [comment]
In addition to the VolumeHost cleanup mentioned above, we need to clean up the volume initialization code (e.g. cmd/kube-controller-manager/app/plugins.go). Consider a putting all plugins into one package at the very least.

**Comment by thockin** [comment]
to clarify the last point for posterity: consider making an all_plugins pkg that just has one function which returns all of the plugins, rather than open-coding the list of them in multiple places, like we are now.

**Comment by saad-ali** [comment]
So three big refactors we still need to do:
1) `volume.VolumePluginMgr` which is required to `InitPlugins` requires a `VolumeHost` object which is very kubelet specific (gets `KubeClient`, `VolumeDir`, etc.) so these methods on the attach/detach controller, PV/PVC, etc. return dummy values.
2)  We have two layers of volume interfaces: `VolumePlugin` in [pkg/volume/plugins.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go) and `Mounter`, `Attacher`, etc. in [pkg/volume/volume.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/volume.go).
    * The original intent was that the inner interface would exist for the lifecycle of a volume operation, and the outer interface would exist for the lifecycle of the plugin.
    *  We need to 1) audit the methods and make sure they are in the correct place, 2) make sure that all objects that implement this interface follow dependency injection design patter to make unit testing possible (lots of existing implementations will for example create a client in a Mount call, for example, making it difficult to test).
3) Volume initialization code (e.g. `cmd/kube-controller-manager/app/plugins.go`) is duplicated for every binary. Consider a putting all plugins into one package at the very least.

**Comment by thockin** [comment]
@msau42 It's been a looooong time - is this still a useful idea?

**Comment by aimuz** [comment]
Hello, I was reading through the code and saw a TODO pointing to a problem and wanted to ask if the change has been implemented?

**Comment by thockin** [comment]
@msau42 ping?

**Comment by LeonxLJX** [comment]
I'd like to take this one.

I've read through the report. Before changing anything I'd like to reproduce it against current `main` first — my reasoning is that a fix anchored to the actual observed behavior is much easier to review than one inferred from the report alone, especially if the surrounding code has moved since this was filed.

Plan: reproduce and confirm the root cause, then make the smallest change that addresses it (plus a regression test if the repo has coverage for this area). If the root cause turns out to live somewhere other than where the report points, I'll report back here before forcing a change.

Anything already known about this that I should read first?

(claiming via @LeonxLJX)
